control.go

```diff
@@ -488,6 +499,14 @@ func (c *controlConn) close() {
 		c.quit <- struct{}{}
 	}
 
+	c.mu.Lock()
+	if c.closed {
```

Comment:
The CompareAndSwapInt32 above already checks if `controlConn.close()` was called, why introduce a separate field?

You could achieve the same effect with:

```go
func (c *controlConn) close() {
	if !atomic.CompareAndSwapInt32(&c.started, 1, -1) {
		return
	}
	ch := c.getConn()
	if ch != nil {
		ch.conn.Close()
	}
}
```
(although clearly `c.started` could be named better)